### ISSUE TITLE
[XQ-DC54/pdx235] camera is not working

### ISSUE THREAD

**Issue by PaoloP79** [sonyxperiadev/bug_tracker#815]
**Platform:** XQ-DC54/pdx235
**Device:** Xperia 10V
**Kernel version:** Linux localhost 5.4.210-qgki-g3750a665ef46 #1 SMP PREEMPT Wed Nov 8 09:04:26 CET 2023 aarch64 Toybox
**Android version:** 13
**Software binaries version:**  [SW_binaries_for_Xperia_Android_13_5.4_v2_zambezi.zip](https://developer.sony.com/file/download/software-binaries-for-aosp-android-13-0-kernel-5-4-zambezi-v2)

**Description**
Following the [Build AOSP Android13 instructions](https://developer.sony.com/open-source/aosp-on-xperia-open-devices/guides/aosp-build-instructions/build-aosp-android-13/) I successfully built and flashed Android13.

After the patch published some days ago now it boot, but camera is not working. The adb logcat output shows that there is a segmentation fault of ``/vendor/bin/hw/android.hardware.camera.provider@2.5-service_64``.

**Symptoms**
* Camera app crash
* Wifi Qrcode scan screen is blank

**How to reproduce**
* Follow the [Build AOSP Android13 instructions](https://developer.sony.com/open-source/aosp-on-xperia-open-devices/guides/aosp-build-instructions/build-aosp-android-13/) 
* Open the camera App, or try to scan a wifi qrcode.

**Additional context**
Part of logcat output
```bash
DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
11-09 13:32:29.826  2546  2546 F DEBUG   : Build fingerprint: 'Sony/aosp_xqdc54/pdx235:13/TQ3A.230901.001/eng.paolo.20231108.091059:userdebug/test-keys'
11-09 13:32:29.826  2546  2546 F DEBUG   : Revision: '0'
11-09 13:32:29.826  2546  2546 F DEBUG   : ABI: 'arm64'
11-09 13:32:29.826  2546  2546 F DEBUG   : Timestamp: 2023-11-09 13:32:29.753927472+0100
11-09 13:32:29.826  2546  2546 F DEBUG   : Process uptime: 0s
11-09 13:32:29.826  2546  2546 F DEBUG   : Cmdline: /vendor/bin/hw/android.hardware.camera.provider@2.5-service_64
11-09 13:32:29.826  2546  2546 F DEBUG   : pid: 2517, tid: 2517, name: android.hardwar  >>> /vendor/bin/hw/android.hardware.camera.provider@2.5-service_64 <<<
11-09 13:32:29.826  2546  2546 F DEBUG   : uid: 1047
11-09 13:32:29.826  2546  2546 F DEBUG   : signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0000000000000028
11-09 13:32:29.826  2546  2546 F DEBUG   : Cause: null pointer dereference
11-09 13:32:29.826  2546  2546 F DEBUG   :     x0  0000000000000028  x1  0000000000000000  x2  0000000000000004  x3  0000000000000004
11-09 13:32:29.826  2546  2546 F DEBUG   :     x4  0000007fe086546f  x5  0000007fe08651b8  x6  0000000035373336  x7  0000000035373336
11-09 13:32:29.826  2546  2546 F DEBUG   :     x8  7df6f28dcb0af860  x9  7df6f28dcb0af860  x10 00000000000009d5  x11 00000000000009d5
11-09 13:32:29.826  2546  2546 F DEBUG   :     x12 0000000000000007  x13 79f4c8b841d28709  x14 0000000000000006  x15 ffffffffffffffff
11-09 13:32:29.826  2546  2546 F DEBUG   :     x16 000000799575cc10  x17 00000079956e6660  x18 000000799a63c000  x19 00000076dadeea50
11-09 13:32:29.826  2546  2546 F DEBUG   :     x20 00000076d843c0f0  x21 00000076dadf1710  x22 00000076dadf1848  x23 00000076dadf06c8
11-09 13:32:29.826  2546  2546 F DEBUG   :     x24 0000007fe08656b0  x25 0000000000000000  x26 0000000000000000  x27 000000799a204000
11-09 13:32:29.826  2546  2546 F DEBUG   :     x28 0000007997578970  x29 0000007fe08654a0
11-09 13:32:29.826  2546  2546 F DEBUG   :     lr  00000076d843b710  sp  0000007fe0865450  pc  00000076d843b710  pst 0000000060001000
11-09 13:32:29.826  2546  2546 F DEBUG   : backtrace:
11-09 13:32:29.826  2546  2546 F DEBUG   :       #00 pc 0000000000011710  /odm/lib64/com.qti.chiusecaseselector.so (PopulatePipelineData.cfi+244) (BuildId: ca0402035f52829b3775809d80aa674c)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #01 pc 00000000003ca380  /odm/lib64/hw/com.qti.chi.override.so (ExtensionModule::ExtensionModule()+1972) (BuildId: 4f98c35a5576088ed0e2faf679d1e0bd)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #02 pc 00000000003c56d4  /odm/lib64/hw/com.qti.chi.override.so (ExtensionModule::GetInstance() (.cfi)+48) (BuildId: 4f98c35a5576088ed0e2faf679d1e0bd)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #03 pc 00000000003c5128  /odm/lib64/hw/com.qti.chi.override.so (chi_hal_override_entry.cfi+20) (BuildId: 4f98c35a5576088ed0e2faf679d1e0bd)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #04 pc 00000000007529a4  /odm/lib64/hw/camera.qcom.so (CamX::HAL3Module::HAL3Module()+2532) (BuildId: d880808970374e7b213d76bfe7238669)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #05 pc 0000000000751dec  /odm/lib64/hw/camera.qcom.so (CamX::HAL3Module::GetInstance() (.cfi)+52) (BuildId: d880808970374e7b213d76bfe7238669)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #06 pc 00000000007496ec  /odm/lib64/hw/camera.qcom.so (CamX::get_number_of_cameras() (.cfi)+108) (BuildId: d880808970374e7b213d76bfe7238669)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #07 pc 000000000074fd94  /odm/lib64/hw/camera.qcom.so (CamX::get_number_of_cameras() (.cfi)+92) (BuildId: d880808970374e7b213d76bfe7238669)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #08 pc 000000000002b530  /vendor/lib64/android.hardware.camera.provider@2.4-legacy.so (android::hardware::camera::common::V1_0::helper::CameraModule::init()+192) (BuildId: d440fb8c4f883782e2b9a9363b864513)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #09 pc 000000000001f338  /vendor/lib64/android.hardware.camera.provider@2.4-legacy.so (android::hardware::camera::provider::V2_4::implementation::LegacyCameraProviderImpl_2_4::initialize()+136) (BuildId: d440fb8c4f883782e2b9a9363b864513)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #10 pc 0000000000027cc0  /vendor/lib64/android.hardware.camera.provider@2.4-legacy.so (android::hardware::camera::provider::V2_4::implementation::LegacyCameraProviderImpl_2_4::LegacyCameraProviderImpl_2_4()+160) (BuildId: d440fb8c4f883782e2b9a9363b864513)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #11 pc 00000000000020f0  /vendor/bin/hw/android.hardware.camera.provider@2.5-service_64 (main.cfi+116) (BuildId: 26ec8050781fefd616281ca22b16bb86)
11-09 13:32:29.826  2546  2546 F DEBUG   :       #12 pc 000000000004a064  /apex/com.android.runtime/lib64/bionic/libc.so (__libc_init+96) (BuildId: a233a8c1a23ee63f36f4b138880288f3)
```

**Comment by jerpelea** [comment]
Hi Paolo,
as I mentioned in the other ticket the camera and PFC are still work in progress 
Best regards
Alin

**Comment by christian-nils** [comment]
Hi,

Any update about this issue? Is a fix still a work in progress? Or did it stall?

Thanks

**Comment by jerpelea** [comment]
still work in progress and will continue after the device are ported to the new kernel

**Comment by gtrentalancia** [comment]
This is also happening on the Sony Xperia 10 IV (Murray platform) running Android 14 (v7b ODM binary package)...

I found the following errors in logcat:

 20:19:07.784 11514 11514 E CamX    : [ERROR][CORE   ] camxsettingsmanager.cpp:538 LoadCamxSettingsOps() **Failed to load settings library com.qti.settings.sm8350.so**
 20:19:07.784 11514 11514 E CamX    : [ERROR][CORE   ] camxsettingsmanager.cpp:548 LoadCamxSettingsOps() Fail to load  settingsOps!
 20:19:07.785 11514 11514 F libc    : Fatal signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x0 in tid 11514 (android.hardwar), pid 11514 (android.hardwar) 

However the mentioned library is available in /odm/lib64/com.qti.settings.sm8350.so.

The backtrace on Murray is only slightly different:

 F DEBUG   : backtrace:
 F DEBUG   :       #00 pc 000000000110d888  /odm/lib64/hw/camera.qcom.so (CSLQueryCameraPlatform+132) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #01 pc 0000000000f5cd14  /odm/lib64/hw/camera.qcom.so (CamX::ProbeChiComponents(CamX::ExternalComponentInfo*, unsigned int*)+172) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #02 pc 0000000000fb1e70  /odm/lib64/hw/camera.qcom.so (CamX::HwEnvironment::Initialize()+768) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #03 pc 0000000000fb1acc  /odm/lib64/hw/camera.qcom.so (CamX::HwEnvironment::HwEnvironment()+96) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #04 pc 0000000000fa7388  /odm/lib64/hw/camera.qcom.so (CamX::HwEnvironment::GetInstance() (.cfi)+52) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #05 pc 00000000007cf378  /odm/lib64/hw/camera.qcom.so (CamX::HAL3Module::HAL3Module()+632) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #06 pc 00000000007cef38  /odm/lib64/hw/camera.qcom.so (CamX::HAL3Module::GetInstance()+52) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #07 pc 00000000007c6908  /odm/lib64/hw/camera.qcom.so (CamX::get_number_of_cameras() (.cfi)+108) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #08 pc 00000000007ccf60  /odm/lib64/hw/camera.qcom.so (CamX::get_number_of_cameras() (.cfi)+92) (BuildId: 64164be92566063a1e3248885721757d)
 F DEBUG   :       #09 pc 000000000002842c  /vendor/lib64/android.hardware.camera.provider@2.4-legacy.so (android::hardware::camera::common::helper::CameraModule::init()+192) (BuildId: 33735351b4da03f9350007000c92d894)
 F DEBUG   :       #10 pc 000000000001c2f8  /vendor/lib64/android.hardware.camera.provider@2.4-legacy.so (android::hardware::camera::provider::V2_4::implementation::LegacyCameraProviderImpl_2_4::initialize()+260) (BuildId: 33735351b4da03f9350007000c92d894)
 F DEBUG   :       #11 pc 000000000001b908  /vendor/lib64/android.hardware.camera.provider@2.4-legacy.so (android::hardware::camera::provider::V2_4::implementation::LegacyCameraProviderImpl_2_4::LegacyCameraProviderImpl_2_4()+156) (BuildId: 33735351b4da03f9350007000c92d894)
 F DEBUG   :       #12 pc 0000000000002144  /vendor/bin/hw/android.hardware.camera.provider@2.5-service_64 (main.cfi+116) (BuildId: e4f4636c7488631cba5b19c4b5cc28dd)
 F DEBUG   :       #13 pc 0000000000054658  /apex/com.android.runtime/lib64/bionic/libc.so (__libc_init+116) (BuildId: 098a1fd01856889214afa7e6d825efbc)

I am wondering if a solution to the problem has been found ?

**Comment by gtrentalancia** [comment]
The camera is working fine on Stock ROM, so there must be a solution available...

**Comment by bartcubbins** [comment]
> 20:19:07.784 11514 11514 E CamX : [ERROR][CORE ] camxsettingsmanager.cpp:538 LoadCamxSettingsOps() **Failed to load settings library com.qti.settings.sm8350.so**

Murray/Zambezi should request `com.qti.settings.mannar.so` instead of `com.qti.settings.sm8350.so`...

**Comment by gtrentalancia** [comment]
Thanks for replying Pavel !

I have searched the `com.qti.settings.mannar.so` in the OEM binary files distributed by Sony at [https://opendevices.sony.net/aosp-on-xperia-open-devices/downloads/software-binaries](https://opendevices.sony.net/aosp-on-xperia-open-devices/downloads/software-binaries), but unfortunately such library is missing !

Where did you get that file from ?

Is there anything else that needs to be changed ?

**Comment by gtrentalancia** [comment]
The only version I could find in the Sony archives of `com.qti.settings.mannar.so` (and `com.qti.feature2.gs.mannar.so`) dates back to the Android 13 v4b software binary package.

However, after installing the above mentioned libraries from the Android 13 package, I get the following errors in logcat:

09-21 18:59:48.254  4008  4008 V CamX    : [ VERB][CORE   ] camxsettingsmanager.cpp:529 LoadCamxSettingsOps() Loading from:com.qti.settings.mannar.so, size:56
09-21 18:59:48.255     0     0 I CAM_INFO: CAM-SMMU: cam_smmu_need_force_alloc_cached: 383 force_cache_allocs=0
09-21 18:59:48.264  4008  4008 E DEWARP  : [ERROR  ] camxchinodedewarp.cpp:2297 ChiNodeEntry() Dewarp Node OpenGL CHINODEEntry funtion
09-21 18:59:48.277  4008  4008 E CamX    : [ERROR][UTILS  ] camxosutilslinux.cpp:1085 LibMap() dlopen: dlopen failed: cannot locate symbol "__emutls_get_address" referenced by "/odm/lib64/com.qti.stats.common.so"..., on '/odm/lib64/camera/components/com.qti.stats.aec.so'
09-21 18:59:48.279  4008  4008 E CamX    : [ERROR][UTILS  ] camxosutilslinux.cpp:1085 LibMap() dlopen: dlopen failed: cannot locate symbol "__emutls_get_address" referenced by "/odm/lib64/com.qti.stats.common.so"..., on '/odm/lib64/camera/components/com.qti.stats.aec.so'
09-21 18:59:48.280  4008  4008 E CamX    : [ERROR][UTILS  ] camxosutilslinux.cpp:1085 LibMap() dlopen: dlopen failed: cannot locate symbol "__emutls_get_address" referenced by "/odm/lib64/com.qti.stats.common.so"..., on '/odm/lib64/camera/components/com.qti.stats.awb.so'
09-21 18:59:48.282  4008  4008 E CamX    : [ERROR][UTILS  ] camxosutilslinux.cpp:1085 LibMap() dlopen: dlopen failed: cannot locate symbol "__emutls_get_address" referenced by "/odm/lib64/com.qti.stats.common.so"..., on '/odm/lib64/camera/components/com.qti.stats.awb.so'
09-21 18:59:48.282  4008  4008 E CamX    : [ERROR][CHI    ] camxchicomponent.cpp:271 GetAWBAlgorithmCallback() Failed to retrieve ChiAWBAlgorithmEntry function address. from lib:/odm/lib64/camera/components/com.qti.stats.awb.so
09-21 18:59:48.287     0     0 I sm5038_fg_check_reg_init_need: SM5038_REG_FG_SYSTEM_STATUS : 0x1010 , return FALSE NO init need
09-21 18:59:48.295  4008  4008 E android.hardware.camera.provider@2.5-service_64: ChiAWBAlgorithmEntry
09-21 18:59:48.296  4008  4008 E CamX    : [ERROR][UTILS  ] camxosutilslinux.cpp:1085 LibMap() dlopen: dlopen failed: cannot locate symbol "__emutls_get_address" referenced by "/odm/lib64/com.qti.stats.common.so"..., on '/odm/lib64/camera/components/com.qti.stats.awb.so'
09-21 18:59:48.297  4008  4008 F libc    : Fatal signal 11 (SIGSEGV), code 2 (SEGV_ACCERR), fault addr 0x6f54d68000 in tid 4008 (android.hardwar), pid 4008 (android.hardwar)

So, basically when using the old binaries, there is a missing symbol which causes Segmentation Fault.

Where do I get updated libraries without missing symbols ?

**Comment by gtrentalancia** [comment]
The missing symbol error with old binaries is probably caused by: [https://android-review.googlesource.com/c/toolchain/llvm/+/631496](https://android-review.googlesource.com/c/toolchain/llvm/+/631496).

I suppose updated libraries `com.qti.settings.mannar.so` (and `com.qti.feature2.gs.mannar.so`) linked with NDK>=r18 are needed...

**Comment by gtrentalancia** [comment]
The following libraries might be involved in the missing ___emutls_get_address_ symbol:

`/odm/lib64/camera/components/com.qti.stats.haf.so`
`/odm/lib64/camera/components/com.qti.stats.af.so`
`/odm/lib64/com.qti.stats.common.so`

**Comment by bartcubbins** [comment]
https://opendevices.sony.net/file/download/software-binaries-for-aosp-android-14-0-kernel-5-4-zambezi-v7a
This ODM package most likely contains everything needed, including the fixed `__emutls_get_address`. But the download link is broken

**Comment by jerpelea** [comment]
> https://opendevices.sony.net/file/download/software-binaries-for-aosp-android-14-0-kernel-5-4-zambezi-v7a This ODM package most likely contains everything needed, including the fixed `__emutls_get_address`. But the download link is broken

we are fixing the link ASAP
Thanks for reporting it 



> The missing symbol error might be caused by: https://android-review.googlesource.com/c/toolchain/llvm/+/631496.
> 
> I suppose updated libraries linked with the appropriate NDK release are needed...

Unfortunately the BSP did not fix the issue so we have to fix it in Android side to be able to load the libraries

**Comment by bartcubbins** [comment]
Unfortunately, I don't have the opportunity right now to go back to Android 14/15 to investigate this,
as I'm fully focused on 16/17 in my spare time. However, I distinctly remember the camera working on
one of the ODM releases, because I was working on the camera regulator driver for Murray/Zambezi.

**Comment by bartcubbins** [comment]
The only workaround I can suggest you at this moment is this:
Download this library: https://android.googlesource.com/platform/prebuilts/ndk/+/refs/tags/android-14.0.0_r74/current/sources/cxx-stl/llvm-libc++/libs/arm64-v8a/libc++_shared.so
Flash `SW_binaries_for_Xperia_Android_14_5.4_murray_v6a.img`
Then do this:
```
adb root 
adb remount /odm
adb push libc++_shared.so /odm/lib64/
adb pull /odm/lib64/com.qti.stats.common.so
patchelf --add-needed libc++_shared.so com.qti.stats.common.so
adb push com.qti.stats.common.so /odm/lib64/
adb reboot
```

Most likely, not all sensors will work on this ODM release, but the front and back wide(?) cameras definitely should.

**Comment by jerpelea** [comment]
Hi 

The archive has been restored 
sorry for the inconvenience

Best regards
Alin

**Comment by gtrentalancia** [comment]
I finally managed to resolve the camera issue, thanks to the tips from Pavel !

First I extracted the two shared libraries `com.qti.settings.mannar.so` and `com.qti.feature2.gs.mannar.so` from the Sony Software Binaries package named _SW_binaries_for_Xperia_Android_14_5.4_murray_v6a.zip_ downloaded from [https://opendevices.sony.net/aosp-on-xperia-open-devices/downloads/software-binaries](https://opendevices.sony.net/aosp-on-xperia-open-devices/downloads/software-binaries) and I have placed the two new files in /odm/lib64.

Then I followed the second Pavel's comment [https://github.com/sonyxperiadev/bug_tracker/issues/815#issuecomment-5775002748](https://github.com/sonyxperiadev/bug_tracker/issues/815#issuecomment-5775002748) and resolved the additional issue [https://github.com/sonyxperiadev/bug_tracker/issues/815#issuecomment-5765922478](https://github.com/sonyxperiadev/bug_tracker/issues/815#issuecomment-5765922478) of the missing `__emutls_get_address` symbol provided by the `libc++_shared.so` library from the Android NDK.

Ideally Sony should create a new updated Software Binary package with the two missing files from step1, with the missing `libc++_shared.so` file from step2 and with the QTI Stats libraries patched with _patchelf_ or preferably rebuilt with an appropriate Android NDK release which doesn't require `libc++_shared.so` and ELF patching !

Alin, can you please look after creating a new updated Software Binaries OEM package ?
